define_iterator not using passed begin/end

define_iterator has two arguments for member function pointers to get the beginning respectively end of an iterator. They don't do anything though.

The pointers are passed into the `Iterator_Impl` and held as members but not ever used. Instead, the `Iterator_Impl` always calls 
```
    Iterator_T it = obj->begin();
    Iterator_T end = obj->end();
```
making the originally passed pointers utterly useless.
This is a problem when one needs multiple iterators on the same entity or to wrap a pre-existing data structure which has non-standard begin/end functions.

e.g.

```
    Data_Type<pkgCache> rb_cPkgCache =
      define_class<pkgCache>("PkgCache")
      .define_iterator(&pkgCache::PkgBegin, &pkgCache::PkgEnd);
```
results in build error
```
/home/me/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/rice-2.1.1/ruby/lib/include/rice/detail/Iterator.hpp:51:32: error: ‘class pkgCache’ has no member named ‘begin’
     Iterator_T it = obj->begin();
                                ^
/home/me/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/rice-2.1.1/ruby/lib/include/rice/detail/Iterator.hpp:52:31: error: ‘class pkgCache’ has no member named ‘end’
     Iterator_T end = obj->end();
                               ^
```
as the `T` in question actually doesn't implement begin/end.

Currently, the only way to deal with this seems to be manually implementing a `.each` or stacking a proxy class on top. Latter still has the problem that one cannot have a second iterator.

--- a/rice/detail/Iterator.hpp
+++ b/rice/detail/Iterator.hpp
@@ -48,8 +48,8 @@ public:
   virtual VALUE call_impl(VALUE self)
   {
     Data_Object<T> obj(self, data_type_);
-    Iterator_T it = obj->begin();
-    Iterator_T end = obj->end();
+    Iterator_T it = (*obj.*begin_)();
+    Iterator_T end = (*obj.*end_)();
     for(; it != end; ++it)
     {
       Rice::protect(rb_yield, to_ruby(*it));

--- a/test/test_Class.cpp
+++ b/test/test_Class.cpp
@@ -303,8 +303,10 @@ public:
   {
   }
 
-  int * begin() { return array_; }
-  int * end() { return array_ + length_; }
+  // Custom names to make sure we call the function pointers rather than
+  // expectable default names.
+  int * beginFoo() { return array_; }
+  int * endBar() { return array_ + length_; }
 
 private:
   int * array_;
@@ -325,7 +327,7 @@ TESTCASE(define_iterator)
 {
   int array[] = { 1, 2, 3 };
   Class c(anonymous_class());
-  c.define_iterator(&Container::begin, &Container::end);
+  c.define_iterator(&Container::beginFoo, &Container::endBar);
   Container * container = new Container(array, 3);
   Object wrapped_container = Data_Wrap_Struct(
       c, 0, Default_Free_Function<Container>::free, container);

